enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,308 @@
+---
+title: an-slo-for-baremetal
+authors:
+  - "@markmc"
+reviewers:
+  - "@abhinavdahiya"
+  - "@dhellmann"
+  - "@enxebre"
+  - "@eparis"
+  - "@hardys"
+  - "@sadasu"
+  - "@smarterclayton"
+  - "@stbenjam"
+approvers:
+  - TBD
+creation-date: 2020-02-13
+last-updated: 2020-02-21
+status: provisional
+see-also:
+ - https://github.com/markmc/cluster-baremetal-operator
+replaces:
+superseded-by:
+---
+
+# An SLO for baremetal
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+1. How to handle upgrades?
+2. Which release should this change target?
+
+## Summary
+
+The Bare Metal Operator provides bare metal machine management
+capabilities needed for the Machine API provider on the `baremetal`
+platform, and there is no equivalent for this component on other
+platforms.
+
+The Bare Metal Operator is currently managed by the Machine API
+Operator in a fashion that requires the Machine API Operator to have
+significant bare metal specific knowledge. This proposal outlines a
+plan for the Bare Metal Operator to become a fully-fledged Second
+Level Operator under the management of the Cluster Version Operator.
+
+(To avoid likely confusion, please note that this proposal describes a
+new project tentatively named Bare Metal Operator (BMO), implying that
+the current baremetal-operator project will be renamed to reflect the
+fact it is "just a controller").
+
```

Comment:
How about naming the new operator "metal3-operator" making it clear that: a. it manages more than the current baremetal-operator and b. it has nothing to do with Baremetal UPI (which does not use metal3)?

Reply:
That does also raise the question if there could be an upstream operator in the metal3-io community repo with an openshift fork, similar to other metal3 components?

Reply:
Naming - in terms of what name shows up in `oc get clusteroperator`, I think that should be a name reflecting the functionality in user terms rather than the software project brand. And if `baremetal` is the name of the clusteroperator, then I think it makes sense to follow that through and metal3 is an implementation detail.

Scope - if we imagine other bare metal related functionality in OpenShift that isn't directly related to the Metal3 project, do we think that should fall under another SLO, or this one? I think it's best to say this new SLO is where bare metal related functionality would be managed.

Upstream project - you could imagine an upstream project which would encapsulate the [kustomize-based deployment scenarios](https://github.com/openshift/baremetal-operator/blob/master/docs/ironic-endpoint-keepalived-configuration.md#kustomization-structure) in the metal3/baremetal-operator project. We could re-use something like that, but we would also need to add OpenShift integration downstream - e.g. the clusteroperator, and checking the platform type in the infrastructure resource. Is there an example of another SLO that is derived from an operator that is more generally applicable?

Reply:
Added this discussion to a new section in the doc